File: src/DataScience/MATLAB/iterator.m
function result = iterator
%ITERATOR Traverse a collection through an explicit iteration protocol.
cursor = struct("items", [10 20 30], "index", 1);
visited = zeros(1, 0);

while hasNext(cursor)
    [value, cursor] = nextValue(cursor);
    visited(end + 1) = value; %#ok<AGROW>
end

result = struct("visited", visited, "finished", ~hasNext(cursor));
end

function answer = hasNext(cursor)
answer = cursor.index <= numel(cursor.items);
end

function [value, cursor] = nextValue(cursor)
if ~hasNext(cursor)
    error("Iterator exhausted");
end
value = cursor.items(cursor.index);
cursor.index = cursor.index + 1;
end

File: src/Other/MicroPython/iterator.py
def make_iterator(values):
    index = 0

    def next_value():
        nonlocal index
        if index >= len(values):
            return None
        current = values[index]
        index += 1
        return current

    return next_value


next_value = make_iterator((10, 20, 30))
visited = []
while True:
    value = next_value()
    if value is None:
        break
    visited.append(value)

if visited != [10, 20, 30] or next_value() is not None:
    raise RuntimeError("Iterator contract failed")
print("iterator=10,20,30")
